derive_state_machine_future/src/phases.rs

```diff
@@ -304,6 +304,298 @@ impl Pass for ValidPaths {
+    type StateMachineExtra = <ValidPaths as Phase>::StateMachineExtra;
+    type StateExtra = BuildStateGenericsExtra;
+}
+
+impl Pass for BuildStateGenerics {
+    type FromPhase = ValidPaths;
+
+    fn pass(machine: StateMachine<ValidPaths>) -> StateMachine<BuildStateGenerics> {
+        // returns a Vec that contains (generic_param, bound_ident, idents).
+        // The bound_ident is the ident of the generic_param (can also be a list of idents).
+        // The idents are all idents that will be collected with the CollectIdents trait.
+        fn prepare_generic_params<'a, T, F, R>(
+            param: &'a [T],
+            get_bound: F,
+        ) -> Vec<(&'a T, R, HashSet<syn::Ident>)>
+        where
+            F: Fn(&'a T) -> R,
+            T: CollectIdents,
+        {
+            param
+                .iter()
+                .map(|p| {
+                    let mut idents = HashSet::new();
+                    p.collect_idents(&mut idents);
+
+                    (p, get_bound(p), idents)
+                })
+                .collect::<Vec<_>>()
+        }
+
+        // Checks if a generic_param is part of the state_idents by calling the contains_bound
+        // function. If contains_bound returns true, this function adds all idents that were
+        // collected for this generic_param. For example <T: Clone>, adds ident 'Clone' to
+        // state_idents. Or, <C, T = C> for T, adds ident 'C' to state_idents.
+        fn extend_state_idents<R, S, F>(
+            state_idents: &mut HashSet<syn::Ident>,
+            param: &[(R, S, HashSet<syn::Ident>)],
+            contains_bound: F,
+        ) where
+            F: Fn(&HashSet<syn::Ident>, &S) -> bool,
+        {
+            let mut iter = param.iter();
+            while let Some(&(_, ref bound, ref idents)) = iter.next() {
+                if contains_bound(state_idents, bound) {
+                    let old_len = state_idents.len();
+                    state_idents.extend(idents.iter().cloned());
+
+                    // When new elements are added, start from the beginning.
+                    // The current generic_param could have added an ident that belongs to a
+                    // generic_param before this generic_param. So, by starting at the beginning,
+                    // we ensure that all generic_params are added.
+                    if old_len != state_idents.len() {
+                        iter = param.iter();
+                        continue;
+                    }
+                }
+            }
+        }
+
+        machine.and_then(|machine, extra, states| {
+            let states = {
+                let mgenerics = &machine.generics;
+
+                // we begin with preparing all params of the machine generics
+                let ty_params = prepare_generic_params(&mgenerics.ty_params, |t| &t.ident);
+                let lifetimes = prepare_generic_params(&mgenerics.lifetimes, |l| &l.lifetime.ident);
```

Comment:
Ahhh OK this was generic so we can treat type parameters and lifetimes separately. That's awesome. We should document that in the `prepare_generic_params` function's comment, because it wasn't clear to me why we had that complexity there.